lib/common/utils.c

```diff
@@ -1738,109 +1738,282 @@ stonith_ipc_server_init(qb_ipcs_service_t **ipcs, struct qb_ipcs_service_handler
+                       const char *user_name, int options, xmlNode **reply)
+{
+    int rc = -ENOTCONN;
+    int max = 5;
+    xmlNode *xml;
+    static enum crm_ipc_flags flags = crm_ipc_flags_none;
+    static gboolean connected = TRUE;
+    static crm_ipc_t *local_ipc = NULL;
+
+    if (reply) {
+        *reply = NULL;
+    }
 
-    crm_xml_add(update, F_ATTRD_VALUE, value);
-    crm_xml_add(update, F_ATTRD_DAMPEN, dampen);
-    crm_xml_add(update, F_ATTRD_SECTION, section);
-    crm_xml_add(update, F_ATTRD_HOST, host);
-    crm_xml_add(update, F_ATTRD_SET, set);
-    crm_xml_add_int(update, F_ATTRD_IS_REMOTE, is_set(options, attrd_opt_remote));
-    crm_xml_add_int(update, F_ATTRD_IS_PRIVATE, is_set(options, attrd_opt_private));
-#if ENABLE_ACL
-    if (user_name) {
-        crm_xml_add(update, F_ATTRD_USER, user_name);
+    /* If caller doesn't provide a connection, use a local one */
+    if (ipc == NULL) {
+        if (local_ipc == NULL) {
+            local_ipc = crm_ipc_new(T_ATTRD, 0);
+            flags |= crm_ipc_client_response;
+            connected = FALSE;
+        }
+        ipc = local_ipc;
     }
-#endif
 
+    /* Build XML to send to attrd */
+    xml = build_attrd_request(command, host, name, value, section, set, dampen,
+                              user_name, options);
+    if (xml == NULL) {
+        return -ENOMEM;
+    }
+
+    /* If we created the connection, we'll try to send up to max times,
+     * otherwise we will only try once
+     */
     while (max > 0) {
         if (connected == FALSE) {
             crm_info("Connecting to cluster... %d retries remaining", max);
             connected = crm_ipc_connect(ipc);
         }
-
         if (connected) {
-            rc = crm_ipc_send(ipc, update, flags, 0, NULL);
+            rc = crm_ipc_send(ipc, xml, flags, 0, reply);
         }
-
-        if (ipc != local_ipc) {
+        if ((ipc != local_ipc) || (rc > 0)) {
             break;
+        }
+        if ((rc != -EAGAIN) && (rc != -EALREADY)) {
+            crm_ipc_close(ipc);
+            connected = FALSE;
+        }
+        sleep(5 - max--);
+    }
+    free_xml(xml);
+    return(rc);
+}
 
-        } else if (rc > 0) {
-            break;
+/*!
+ * \brief Submit a query request to attrd and wait for reply
+ *
+ * \param[in] ipc       Open connection to attrd (or NULL to create local connection)
+ * \param[in] host      If specified, request applies to this host only
+ * \param[in] name      Name of attribute to be affected by request
+ * \param[in] user_name If ACLs are in use and this is specified, user to submit request as
+ * \param[out] values   New hash table of host-value pairs
+ *
+ * \return pcmk_ok on success,
+ *         -errno on error (-ENXIO = requested attribute does not exist)
+ * \note On success, the caller is responsible for freeing the new hash table
+ *       using g_hash_table_destroy().
+ */
+int
+attrd_query_delegate(crm_ipc_t *ipc, const char *host, const char *name,
```

Comment:
I put it in utils.c so attrd_request_delegate() could be static. Would you prefer for me to (1) expose attrd_request_delegate() internally and move attrd_query_delegate() to attrd_updater.c, or (2) expose attrd_query_delegate() internally rather than publicly?